src/tree/tree_model.cc

```diff
@@ -618,6 +622,165 @@ std::string RegTree::DumpModel(const FeatureMap& fmap,
+    GetChar('[');
+    for (int32_t i = 0; i < n_nodes; ++i) {
+      GetChar('[');
+      auto& s = stats[i];
+      size_t pos {0};
+      s.loss_chg = std::stof(raw_str_.substr(cursor_.Pos(), kMaxNumLength), &pos);
+      ForwardN(pos);
+      GetChar(',');
+      s.sum_hess = std::stof(raw_str_.substr(cursor_.Pos(), kMaxNumLength), &pos);
+      ForwardN(pos);
+      GetChar(',');
+      s.base_weight = std::stof(raw_str_.substr(cursor_.Pos(), kMaxNumLength), &pos);
+      ForwardN(pos);
+      GetChar(',');
+      s.leaf_child_cnt = std::stoi(raw_str_.substr(cursor_.Pos(), kMaxNumLength), &pos);
+      ForwardN(pos);
+      GetChar(']');
+      if (i != n_nodes - 1) {
+        GetChar(',');
+      }
+    }
+    GetChar(']');
+  }
+
+  void ParseNode(int32_t n_nodes, std::vector<Node>* p_nodes) {
+    p_nodes->resize(n_nodes);
+    auto& nodes = *p_nodes;
+    GetChar('[');
+    for (int32_t i = 0; i < n_nodes; ++i) {
+      GetChar('[');
+      size_t pos {0};
+      auto left = std::stoi(raw_str_.substr(cursor_.Pos(), kMaxNumLength), &pos);
+      ForwardN(pos);
+      GetChar(',');
+      auto right = std::stoi(raw_str_.substr(cursor_.Pos(), kMaxNumLength), &pos);
+      ForwardN(pos);
+      GetChar(',');
+      auto par = std::stoi(raw_str_.substr(cursor_.Pos(), kMaxNumLength), &pos);
+      ForwardN(pos);
+      GetChar(',');
+
+      auto split = common::stoui(raw_str_.substr(cursor_.Pos(), kMaxNumLength), &pos);
+      ForwardN(pos);
+      GetChar(',');
+
+      auto cond = std::stof(raw_str_.substr(cursor_.Pos(), kMaxNumLength), &pos);
+      ForwardN(pos);
+      GetChar(',');
+
+      auto default_left = common::stoui(raw_str_.substr(cursor_.Pos(),
+                                                        kMaxNumLength), &pos);
+      ForwardN(pos);
+      GetChar(']');
+      nodes[i]= {left, right, par, split, cond, static_cast<bool>(default_left)};
+      if (i != n_nodes - 1) {
+        GetChar(',');
+      }
+    }
+    GetChar(']');
+  }
+};
+
+void RegTree::Load(Json const& in) {
+  param.InitAllowUnknown(fromJson(get<Object>(in["tree_param"])));
+  {
+    auto stats_in = get<Raw>(in["stats"]);
+    TreeReader reader(StringView{stats_in.c_str(), stats_in.size()});
+    reader.ParseStat(param.num_nodes, &stats_);
+    CHECK_EQ(param.num_nodes, stats_.size());
+  }
+
+  {
+    auto nodes_in = get<Raw>(in["nodes"]);
+    TreeReader reader(StringView{nodes_in.c_str(), nodes_in.size()});
+    reader.ParseNode(param.num_nodes, &nodes_);
+    CHECK_EQ(param.num_nodes, nodes_.size());
+  }
+}
+
+void RegTree::Save(Json* p_out) const {
+  auto& out = *p_out;
+  CHECK_EQ(param.num_nodes, static_cast<int>(nodes_.size()));
+  CHECK_EQ(param.num_nodes, static_cast<int>(stats_.size()));
+  out["tree_param"] = toJson(param);
+  CHECK_EQ(get<String>(out["tree_param"]["num_nodes"]), std::to_string(param.num_nodes));
+
+  std::string tree_raw;
+  tree_raw.reserve(
+      stats_.size() * std::numeric_limits<Number::Float>::max_digits10 * 2);
```

Comment:
I am trying to figure out why you reserve this amount of memory.
It seems each stat has 3 floats and one integer so I would expect that the total length needed is

stats_.size() * ( 2 + float_length*3 + int_length + 3) + 2
2 is []
3 are commas
float_length = std::numeric_limits<Number::Float>::max_digits10 + 1
int_length ...